### ISSUE TITLE
dune runtest fails on OCaml 4.12.0

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/dune#3766]
I ran opam-health-check on 4.12 overnight and I'm seeing lots of failures on OCaml 4.12.0 that all seem to be related to dune:
```
#=== ERROR while compiling re.1.9.0 ===========================================#
# context              2.1.0~beta2 | linux/x86_64 | ocaml-variants.4.12.0+trunk | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.0+trunk/.opam-switch/build/re.1.9.0
# command              ~/.opam/4.12.0+trunk/bin/dune runtest -p re -j 71
# exit-code            1
# env-file             ~/.opam/log/re-1802-b1a82b.env
# output-file          ~/.opam/log/re-1802-b1a82b.out
### output ###
# File "lib_test/dune", line 4, characters 47-56:
# 4 |  (names test_easy test_re test_perl test_emacs test_glob test_pcre))
#                                                    ^^^^^^^^^
# Error: File unavailable:
# /home/opam/.opam/4.12.0+trunk/lib/stdlib-shims/stdlib_shims.a
```
They all seem to fail when `dune runtest` or `dune build @runtest` is called and only appears on 4.12.0.
https://github.com/ocaml/dune/pull/3576 and https://github.com/ocaml/opam-repository/pull/16983 seem to be related issues (cc @dra27)

## Reproduction

1. Install ocaml 4.12.0+trunk
2. call `opam install -t re.1.9.0`

## Specifications

- Version of `dune` (output of `dune --version`): 2.7.0

**Comment by kit-ty-kate** [comment]
Dune also fails on `dune build` when an executable is present it seems:
```
41/42: [tuntap: dune build] + /home/opam/.opam/4.12.0+trunk/bin/dune "build" "-p" "tuntap" "-j" "71" (CWD=/home/opam/.opam/4.12.0+trunk/.opam-switch/build/tuntap.2.0.0) - File "bin/dune", line 2, characters 7-14: - 2 | (name otunctl) - ^^^^^^^ - Error: File unavailable: - /home/opam/.opam/4.12.0+trunk/lib/stdlib-shims/stdlib_shims.a
```

**Comment by kit-ty-kate** [comment]
for the record this also happens with dune 2.7.1

**Comment by rgrinberg** [comment]
We've discussed this on slack and reached the following conclusion:

* Libraries without any modules should specify this with an empty `(modules)` field.
* Empty libraries without a `(modules)` field will be rejected for dune >= 2.8 && ocaml >= 4.12 (since dune does not build such libraries correctly without a `(modules) field
* Empty libraries without a `(modules)` field but with ocaml < 4.12 will trigger a warning.

**Comment by ejgallego** [comment]
I'm seeing this problem, unfortunately adding `(modules)` doesn't workaround the problem.

**Comment by kit-ty-kate** [comment]
Fixed in #3973. Thanks a lot!
